File: 1260-shift-2d-grid/1260-shift-2d-grid.py
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        r, c = len(grid), len(grid[0])
        n = r * c
        k = k % n

        if not k: return grid

        rev(grid, 0, n - 1)
        rev(grid, 0, k - 1)
        rev(grid, k, n - 1)

        return grid

def rev(grid, i, j):
    c = len(grid[0])
    while i < j:
        grid[i // c][i % c], grid[j // c][j % c] = grid[j // c][j % c], grid[i // c][i % c]
        i += 1
        j -= 1

# array, simulation
# time: O(n)
# space: O(1)

# this approach is (prolly) the most optimal one

File: 1260-shift-2d-grid/1260-shift-2d-grid.go
func shiftGrid(grid [][]int, k int) [][]int {
	r, c := len(grid), len(grid[0])
	n := r * c
	k = k % n

	if k == 0 {
		return grid
	}

	temp := make([]int, n)

	for i := 0; i < r; i++ {
		for j := 0; j < c; j++ {
			temp[(i*c)+j] = grid[i][j]
		}
	}

    temp = append(temp[len(temp)-k:], temp[:len(temp)-k]...)

    for i := 0; i < len(temp); i++ {
        grid[i/c][i%c] = temp[i]
    }

    return grid
}

// array
// time: O(n)
// space: O(n)